src/components/tl/mlx5/mcast/tl_mlx5_mcast_helper.h

```diff
@@ -0,0 +1,322 @@
+            tl_error(comm->lib, "post send failed: ret %d, start_psn %d, to_send %d, "
+                      "to_recv %d, length %d, psn %d, inline %d",
+                      rc, req->start_psn, req->to_send, req->to_recv,
+                      length, pp->psn, length <= comm->max_inline);
+            return UCC_ERR_NO_MESSAGE;
+        }
+
+        status = ucc_tl_mlx5_mcast_check_nack_requests(comm, pp->psn);
+        if (UCC_OK != status) {
+            return status;
+        }
+    }
+
+    req->offset = offset;
+
+    return UCC_OK;
+}
+
+/* this function return the number of mcast recv packets that
+ * have been arrived or -1 in case of error */
+static inline int ucc_tl_mlx5_mcast_recv(ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                         ucc_tl_mlx5_mcast_coll_req_t *req,
+                                         int num_left, int *pending_q_size)
+{
+    struct pp_packet *pp;
+    struct pp_packet *next;
+    uint64_t          id;
+    struct ibv_wc    *wc;
+    int               num_comp;
+    int               i;
+    int               real_num_comp;
+
+    /* check if we have already received something */
+    ucc_list_for_each_safe(pp, next, &comm->pending_q, super) {
+        if (PSN_IS_IN_RANGE(pp->psn, req, comm)) {
+            __builtin_prefetch(req->ptr + PSN_TO_RECV_OFFSET(pp->psn, req, comm));
+            __builtin_prefetch((void*) pp->buf);
+            ucc_list_del(&pp->super);
+            ucc_tl_mlx5_mcast_process_packet(comm, req, pp);
+            num_left--;
+        } else if (pp->psn < comm->last_acked){
+            ucc_list_del(&pp->super);
+            ucc_list_add_tail(&comm->bpool, &pp->super);
+        }
+
+        (*pending_q_size)++;
+    };
+
+    wc = ucc_malloc(sizeof(struct ibv_wc) * POLL_PACKED, "WC");
+    if (!wc) {
+        tl_error(comm->lib, "ucc_malloc failed");
+        return -1;
+    }
+
+    while (num_left > 0)
+    {
+        memset(wc, 0, sizeof(struct ibv_wc) * POLL_PACKED);
+        num_comp = ibv_poll_cq(comm->rcq, POLL_PACKED, wc);
+
+        if (num_comp < 0) {
+            tl_error(comm->lib, "recv queue poll completion failed %d", num_comp);
+            ucc_free(wc);
+            return -1;
+        } else if (num_comp == 0) {
+            break;
+        }
+
+        real_num_comp = num_comp;
+
+        for (i = 0; i < real_num_comp; i++) {
+            if (IBV_WC_SUCCESS != wc[i].status) {
+                tl_error(comm->lib, "mcast_recv: %s err pending_recv %d wr_id %ld"
+                         " num_comp %d byte_len %d",
+                         ibv_wc_status_str(wc[i].status), comm->pending_recv,
+                         wc[i].wr_id, num_comp, wc[i].byte_len);
+                ucc_free(wc);
+                return -1;
+            }
+
+            id         = wc[i].wr_id;
+            pp         = (struct pp_packet*) (id);
+            pp->length = wc[i].byte_len - GRH_LENGTH;
```

Comment:
do you need to check for IBV_WC_GRH?

Reply:
As of now, we do not need the content of GRH field of the received UD packet. Therefore, we do not need to check whether or not the GRH field is valid.